### PR TITLE
feat: [PLG-1408]: Added Stripe sync event to finalize invoice after avalara…

### PR DESCRIPTION
**Pull request by thomas-grasman**
… tax is added

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/36175)

### CODE REVIEW

**Comment by diptiman-adak** [comment]
Quality gate Quality gate: **_Failed_** 
 


            
 #### 1 Issues 
* Code issue 1 Code issues 

 [Details](https://app.embold.io/organization/gh/harness/repositories/7752dbb3099d2dba3b5d37f5931a5e4c/pull-request/36175)

**Review by hekimoglu010 — APPROVED** [review]
> Reviewed 5 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/36175)* status: 5 file(s) to review. 1 of 2 approvals obtained.

File: 925-subscription-service/src/main/java/io/harness/subscription/handlers/InvoiceUpdatedHandler.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+  @Inject
+  public InvoiceUpdatedHandler(StripeHelper stripeHelper) {
+    this.stripeHelper = stripeHelper;
+  }
+
+  @Override
+  public void handleEvent(Event event) {
+    Invoice invoice = StripeEventUtils.convertEvent(event, Invoice.class);
+    if (invoice.getLines().getData().stream().anyMatch(n -> n.getMetadata().get("Is_AvaTax_Line").equals("true"))) {
```
Comment (pranil-harness) [inline 925-subscription-service/src/main/java/io/harness/subscription/handlers/InvoiceUpdatedHandler.java:31]: Make "Is_AvaTax_Line" a constant

File: 925-subscription-service/src/main/java/io/harness/subscription/handlers/InvoiceUpdatedHandler.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+      finalizeInvoice(invoice);
+    }
+  }
+
+  private void finalizeInvoice(Invoice invoice) {
+    if (invoice.getPaymentIntent() == null) {
```
Comment (pranil-harness) [inline 925-subscription-service/src/main/java/io/harness/subscription/handlers/InvoiceUpdatedHandler.java:37]: What happens if Invoice is not null? Do we assume it to be the finalized version?

Comment (thomas-grasman) [reply 925-subscription-service/src/main/java/io/harness/subscription/handlers/InvoiceUpdatedHandler.java:39]: Yes. In order for the payment intent to exist, the invoice must be finalized. The only reason we are finalizing them manually is to create this payment intent to retrieve the clientSecret.

Comment (pranil-harness) [reply 925-subscription-service/src/main/java/io/harness/subscription/handlers/InvoiceUpdatedHandler.java:39]: What happens if the finalization fails? You're returning void here? any exception handing to be done?

Comment (thomas-grasman) [reply 925-subscription-service/src/main/java/io/harness/subscription/handlers/InvoiceUpdatedHandler.java:39]: Exception is handled in `public Invoice finalizeInvoice(String invoiceId)`

This is going to occur in a web hook call, so we won't see the results. The exception will be returned to Stripe.

**Review by pranil-harness — APPROVED** [review]

**Review by pranil-harness — COMMENTED** [review]
> Reviewed 4 of 5 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/36175)* status: 1 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by pranil-harness — DISMISSED** [review]

**Review by hekimoglu010 — DISMISSED** [review]
> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/36175)* status: 1. unresolved discussions.

**Review by devinder-harness — APPROVED** [review]

**Review by devinder-harness — APPROVED** [review]

**Review by pranil-harness — COMMENTED** [review]
> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/36175)* status: 1 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by pranil-harness — APPROVED** [review]

**Pull request merged**